File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <stdio.h>
#include <stdlib.h>

void Queen(int row, int n, int* col, int* diag1, int* diag2, int* count) {
	if (row == n) {
		(*count)++;
		return;
	}

	for (int i = 0; i < n; i++) {
		if (!col[i] && !diag1[row + i] && !diag2[row - i + n]) {
			col[i] = diag1[row + i] = diag2[row - i + n] = 1;

			Queen(row + 1, n, col, diag1, diag2, count);

			col[i] = diag1[row + i] = diag2[row - i + n] = 0;
		}
	}
}

int main(void) {
	int n = 0;
	if (scanf("%d", &n) != 1) return 0;

	int* col = (int*)calloc(n, sizeof(int));
	int* diag1 = (int*)calloc((2 * n), sizeof(int));
	int* diag2 = (int*)calloc((2 * n), sizeof(int));

	int count = 0;

	Queen(0, n, col, diag1, diag2, &count);

	printf("%d\n", count);


	free(diag2);
	free(diag1);
	free(col);
	return 0;
}

File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys

input = sys.stdin.readline
write = sys.stdout.write

def Queen(row: int, n: int, col: list[bool], diag1: list[bool], diag2: list[bool], count_list: list[int]):
    global count
    
    if row == n:
        count_list[0] += 1
        return
    
    for i in range(n):
        if not col[i] and not diag1[row + i] and not diag2[row - i + n]:
            col[i] = diag1[row + i] = diag2[row - i + n] = True
            
            Queen(row + 1, n, col, diag1, diag2, count_list)
            
            col[i] = diag1[row + i] = diag2[row - i + n] = False
            
def solve():
    n = int(input().rstrip())
    if not n: return
    
    count_list = [0]
    col = [False] * n
    diag1 = [False] * (2 * n)
    diag2 = [False] * (2 * n)
    
    Queen(0, n, col, diag1, diag2, count_list)
    
    write(str(count_list[0]) + "\n")
    
if __name__ == "__main__":
    solve()